src/contracts/oracles/BeefyVeloVaultRelayer.sol

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import {IBaseOracle} from '@interfaces/oracles/IBaseOracle.sol';
+
+import {IBeefyVeloVaultRelayer} from '@interfaces/oracles/IBeefyVeloVaultRelayer.sol';
+
+import {IBeefyVaultV7} from '@interfaces/external/IBeefyVaultV7.sol';
+import {IVeloPool} from '@interfaces/external/IVeloPool.sol';
+
+import {Math, WAD} from '@libraries/Math.sol';
+
+/**
+ * @title  BeefyVeloVaultRelayer
+ * @notice Deconstructs a Beefy Vault to it's Velodrome liquidity pool and that pool's constituent tokens to return a price feed
+ * @dev    Requires an underlying Velodrome pool and price feeds for the pool's tokens
+ */
+contract BeefyVeloVaultRelayer is IBaseOracle, IBeefyVeloVaultRelayer {
+  using Math for uint256;
+
+  // --- Registry ---
+  /// @inheritdoc IBeefyVeloVaultRelayer
+  IBaseOracle public token0priceSource;
+
+  /// @inheritdoc IBeefyVeloVaultRelayer
+  IBaseOracle public token1priceSource;
+
+  /// @inheritdoc IBeefyVeloVaultRelayer
+  IBeefyVaultV7 public beefyVault;
+
+  /// @inheritdoc IBeefyVeloVaultRelayer
+  IVeloPool public veloPool;
+
+  // --- Data ---
+
+  /// @inheritdoc IBaseOracle
+  string public symbol;
+
+  // --- Init ---
+
+  /**
+   *
+   * @param  _token0priceSource Address of the price source for the first token in the velo pool
+   * @param  _token1priceSource Address of the price source for the second token in the velo pool
+   * @param  _beefyVault The address of the beefy vault contract
+   * @param  _veloPool The address of the velo pool underlying the beefy vault
+   */
+  constructor(
+    IBaseOracle _token0priceSource,
+    IBaseOracle _token1priceSource,
+    IBeefyVaultV7 _beefyVault,
+    IVeloPool _veloPool
+  ) {
+    if (address(_token0priceSource) == address(0)) {
+      revert BeefyVeloVaultRelayer_NullPriceSource();
+    }
+    if (address(_token1priceSource) == address(0)) {
+      revert BeefyVeloVaultRelayer_NullPriceSource();
+    }
+    if (address(_beefyVault) == address(0)) {
+      revert BeefyVeloVaultRelayer_NullBeefyVault();
+    }
+    if (address(_veloPool) == address(0)) {
+      revert BeefyVeloVaultRelayer_NullVeloPool();
+    }
+
+    token0priceSource = _token0priceSource;
+    token1priceSource = _token1priceSource;
+    beefyVault = _beefyVault;
+    veloPool = _veloPool;
+
+    symbol = string(abi.encodePacked(_beefyVault.symbol(), ' / USD '));
+  }
+
+  /// @inheritdoc IBaseOracle
+  function getResultWithValidity() external view returns (uint256 _result, bool _validity) {
+    (uint256 _token0priceSourceValue, bool _token0priceSourceValidity) = token0priceSource.getResultWithValidity();
+
+    (uint256 _token1priceSourceValue, bool _token1priceSourceValidity) = token1priceSource.getResultWithValidity();
+
+    uint256 _totalValue = _getCombinedPriceValue(_token0priceSourceValue, _token1priceSourceValue);
+
+    _validity = _token0priceSourceValidity && _token1priceSourceValidity;
```

Comment:
if 1 is valid but the other is not, you dont want to use the valid one?

Reply:
The quote for the beefy vault moo token should only be considered valid if the quotes for the 2 underlying tokens are both valid.